va/va.go

```diff
@@ -251,12 +251,39 @@ type verificationRequestEvent struct {
 	Error             string `json:",omitempty"`
 }
 
+// ipError is an error type used to pass though the IP address of the remote
+// host when an error occurs during HTTP-01 and TLS-ALPN domain validation.
+type ipError struct {
+	ip  net.IP
+	err error
+}
+
+// Unwrap returns the underlying error.
+func (i ipError) Unwrap() error {
+	return i.err
+}
+
+// Error returns a string representation of the error.
+func (i ipError) Error() string {
+	return fmt.Sprintf("%s: %s", i.ip, i.err)
+}
+
 // detailedError returns a ProblemDetails corresponding to an error
 // that occurred during HTTP-01 or TLS-ALPN domain validation. Specifically it
 // tries to unwrap known Go error types and present something a little more
 // meaningful. It additionally handles `berrors.ConnectionFailure` errors by
 // passing through the detailed message.
 func detailedError(err error) *probs.ProblemDetails {
+	var ipErr ipError
+	if errors.As(err, &ipErr) {
+		detailedErr := detailedError(ipErr.err)
+		if ipErr.ip == nil {
+			return detailedErr
+		}
```

Comment:
This will make some future reader say "wait, when can ipError.ip be nil?" The answer is never (assuming the code is correct) and this is a belt-and-suspenders check to avoid printing `%!s(<nil>)`. Let's add a comment saying "// Should never happen."

We also discussed on a call how we can make _sure_ this never happens. One way would be to use package isolation:

 - put `ipError` in its own package
 - export it as `IPError`
 - export the `.Err` field but only expose the IP via a `.IP()` accessor method
 - export a `New(ip, err) error` constructor
 - in that constructor, if `ip` is nil, pass through `err` without constructing an `IPError`
 - now you can guarantee that the `.IP()` of an `IPError` is never nil!
 
 But this is probably overkill.